source/libs/features.js

```diff
@@ -0,0 +1,187 @@
+import {h} from 'dom-chef';
+import select from 'select-dom';
+import onDomReady from 'dom-loaded';
+import isPromise from 'p-is-promise';
+import OptionsSync from 'webext-options-sync';
+import onNewComments from './on-new-comments';
+import * as pageDetect from './page-detect';
+import {safeElementReady} from './utils';
+
+/*
+ * When navigating back and forth in history, GitHub will preserve the DOM changes;
+ * This means that the old features will still be on the page and don't need to re-run.
+ * For this reason `onAjaxedPages` will only call its callback when a *new* page is loaded.
+ *
+ * Alternatively, use `onAjaxedPagesRaw` if your callback needs to be called at every page
+ * change (e.g. to "unmount" a feature / listener) regardless of of *newness* of the page.
+ */
+async function onAjaxedPagesRaw(callback) {
+	await onDomReady;
+	document.addEventListener('pjax:end', callback);
+	callback();
+}
+function onAjaxedPages(callback) {
+	onAjaxedPagesRaw(() => {
+		if (!select.exists('has-rgh')) {
+			callback();
+		}
+	});
+}
+
+// Must be called after all the features were added to onAjaxedPages
+// to mark the current load as "done", so history.back() won't reapply the same DOM changes.
+// The two `await` ensure this behavior and order.
+onAjaxedPages(async () => {
+	await globalReady; // Match `add()`
+	await Promise.resolve(); // Kicks it to the next tick, after the other features have `run()`
+
+	const ajaxContainer = select('#js-repo-pjax-container,#js-pjax-container');
+	if (ajaxContainer) {
+		ajaxContainer.append(<has-rgh/>);
+	}
+});
+
+// Rule assumes we don't want to leave it pending:
+// eslint-disable-next-line no-async-promise-executor
+const globalReady = new Promise(async resolve => {
+	await safeElementReady('body');
+
+	if (pageDetect.is500()) {
+		return;
+	}
+
+	if (document.body.classList.contains('logged-out')) {
+		console.warn('%cRefined GitHub%c only works when you’re logged in to GitHub.', 'font-weight: bold', '');
+		return;
+	}
+
+	if (select.exists('html.refined-github')) {
+		console.warn('Refined GitHub has been loaded twice. If you didn’t install the developer version, this may be a bug. Please report it to: https://github.com/sindresorhus/refined-github/issues/565');
+		return;
+	}
+
+	document.documentElement.classList.add('refined-github');
+
+	// Options defaults
+	const options = Object.assign(
+		{
+			disabledFeatures: '',
+			customCSS: '',
+			logging: false
+		},
+		await new OptionsSync().getAll()
+	);
+
+	if (options.customCSS.trim().length > 0) {
+		document.head.append(<style>{options.customCSS}</style>);
+	}
+
+	// Create logging function
+	options.log = options.logging ? console.log : () => {};
+
+	resolve(options);
+});
+
+const run = async ({id, include, exclude, init, deinit, options: {log}}) => {
```

Comment:
Not sure whether it's something you want to fix since I guess it means other things are wrong but it looks like this will error with `TypeError: Cannot destructure property `log` of 'undefined' or 'null'.` in some occurances. 

`globalReady` returns `undefined` if refined-github has been loaded twice, you’re logged out, or if it’s a 500 page. This means that `options` is undefined when `run(details)` is called.

Reply:
Did you encounter that? `globalReady` won’t resolve/error at all, it will stay pending in all those cases and no further code will be executed

Reply:
Oh yes, I was thinking it would act like a `.then()` callback and wrap the return value in a resolve. But, no you’re right. thank you